### ISSUE TITLE
Game feels so delayed since the update

### ISSUE THREAD

**Issue by ATTICUSHBL** [haxball/haxball-issues#1895]
Basically this post is made for the simple reason that the game felt different since it suffered the first update.

I guess it is because many options were added, changing the game code, which in turn generated an impact on certain players, increasing input lag. So what i am looking for with this post are at least answers to why is this happening and if it can be solved to improve the game experience, because im not the only one who think this, its more like a popular opinion.

**Comment by HaxPlayer1** [comment]
I totally agree, since the update the token is heavier than before, I don't know exactly if it is the token or the game that is slowing down but certainly there is a noticeable slowness, many friends told me that the same thing is happening to them

**Comment by basro** [comment]
I'd like you to help me investigate this (alleged) performance regression.

I've uploaded a copy of the site before the updates here: https://old-version.haxball.com/play

I ask you to please verify if the site is running worse for you. Try to do it as "scientifically" as possible and try to remove any bias you can. A double blind test would be ideal.

**Comment by HaxPlayer1** [comment]
You really feel the difference, I play identically to how I played before with that level of fluidity, dear basro, do you think you could replicate that same fluidity in a certain way in the new update?
 This one has very positive things like the transparency of the chat and the options in the menu.

**Comment by ATTICUSHBL** [comment]
> 
I aprecciate it a lot.

We were testing it with some friends and we made some comparisons with the "new" version and I was able to form this opinion:

The delay or "input lag" becomes much more noticeable in the new version, I suppose that as I mentioned before it is because new things have been added to the game, so I think that this affects the optimization of the game, and in this case for the worse . Not only do I base this opinion on my thoughts, but it's something we've just been testing and we all came to the same conclusion, so it's not a simple "Crazy" thought.

As a possible suggestion for those of us who are used to how the game was before, not only to the interface but also to the fluidity that the game had before, it would be ideal if this "old" version could be found on the haxball.com page as a simple paragraph at least. 
And it would be nice if you could tell me what do you think caused this impact on the fluidity of the game or why at least some people feel that way? Why did it happened?

Finally I want to thank you on behalf of myself and the entire haxball community in Latin America for all your work in the game and for showing your interest again, I hope that somehow you can solve this problem with the optimization of the new version, we aprecciate all your effort ♡.

**Comment by basro** [comment]
Could you test if the difference in input lag is noticeable if you play with the restricted view mode?

**Comment by ATTICUSHBL** [comment]
> Could you test if the difference in input lag is noticeable if you play with the restricted view mode?

Yes, in fact I play with the restricted view (camera 4 or 0) and that is where the input lag is most noticeable (in the new version), although I have tried camera 3 in both versions with some people and the difference is still there, but maybe it's not something that worries you instantly as in my case the restricted view does.

**Comment by basro** [comment]
Could you tell me your pc specs?

**Comment by ATTICUSHBL** [comment]
> Could you tell me your pc specs?

![image](https://user-images.githubusercontent.com/132158785/235329676-49510622-a72a-4019-9cf5-f6b35cb8e236.png)

**Comment by basro** [comment]
I've been researching this issue and I've made a few changes that I'd like you to test.

**Please test only with the latest version of chrome.** One of the most important changes is making use of a new-ish chrome feature that is not available in some older versions, the feature is also not available in firefox.

The changes are available in this url: https://next.haxball.com/play

Please compare this new version with the one in www.haxball.com and the one in old-version.haxball.com

I'm cautiously optimistic.

**Comment by kingjasko** [comment]
I'm aware it probably has nothing to do with you, basro, or haxball itself, but since some of the recent chrome updates (1st time i noticed it was probably around 2-3 weeks ago), unlock fps command (--disable-frame-rate-limit) completely breaks haxball for me, here's a video of that happening - https://youtu.be/AxaVUr6id0M

I tried the same on https://next.haxball.com/play and I just get limited to 60fps even on my 144hz monitor - https://youtu.be/yDhHej2j8c8

Anyway, is there anything I can do about it? I wouldn't want to go back to outdated chrome versions or experiment even further with other electron/chrome based browsers just for this reason. FYI, I'm definitely not the only one with this issue currently.

**Comment by basro** [comment]
@kingjasko 

I've also been investigating that and was going to be posting about it soon. I did some hacky experiments and was able to make current chrome not freeze. The way I did it wasy extremely hacky and I want to test more robust and elegant ways to do it. Chrome is behaving really weirdly sadly.

Does next.haxball.com also limit itself to 60 fps if you run it in normal chrome without the fps limit flag?

**Comment by kingjasko** [comment]
> Does next.haxball.com also limit itself to 60 fps if you run it in normal chrome without the fps limit flag?

It does not.

**Comment by basro** [comment]
Could you test if you can perceive an improvement in input lag between next and www? (it's not going to be better than running with fps limit disabled, but any improvement is a win here)

**Comment by kingjasko** [comment]
tested the next one here - https://youtu.be/_Wr-IGD2_mI

tested the current one here - https://youtu.be/aYdWMO5CIaw

I honestly don't know if it's placebo or if I felt an actual difference, but to me, it felt better on next.haxball, it just felt more consistent, i immediately gained more confidence when it comes to ball control, double shots and general kicks. By feeling, I'd say, yes, input lag is lower, but take it with a grain of salt. :D

Oh and my specs if it matters ryzen 9 5900x, 32gb of ram, rx 6600 8gb, latest versions of win10 and chrome.

**Comment by basro** [comment]
> I honestly don't know if it's placebo

Well, I'm hoping it's not. I've found a feature in chrome called desynchronized canvas rendering which is intended to remove lag when rendering to canvas. It was released late 2019 but I only found it existed now.

In my own pc I've confirmed that at least on the mouse input lag test I get 30 less milliseconds of input lag (60 hz monitor, 2 frames).

The feature is this: https://developer.chrome.com/blog/desynchronized/

**Comment by basro** [comment]
Unfortunately I'm very bad at noticing input lag when playing haxball which is why I need others to confirm if this is an improvement or not. If possible please get other people to test it, specially if you know they are sensitive to input lag.

**Comment by kingjasko** [comment]
2 other people said they definitely feel a positive difference, so there you go

**Comment by basro** [comment]
Finally something seems to go right

**Comment by kingjasko** [comment]
https://basro.github.io/input-lag-measuring-tool/ are you able to make another url like you did with next.haxball so it uses this new feature? Shouldn't the difference be easily noticeable there?

**Comment by basro** [comment]
https://basro.github.io/input-lag-measuring-tool/ normal canvas
https://basro.github.io/input-lag-measuring-tool/desynchronized_canvas desynchronized canvas

A word of warning: Chrome is weird and after running the desynchronized canvas test switching back to a normal canvas will often also have lower latency. What I've found is that if you open 2 normal canvas test tabs and quickly switch between them chrome will go back to having high latency. When that happens switching to the desynchronized canvas tab always reduces the latency for me.

**Comment by basro** [comment]
Here's a video showing what I mean regarding the tab switching.
https://www.youtube.com/watch?v=zKgyAm_Yog0

**Comment by kingjasko** [comment]
well, there is definitely a measurable difference, it became very clear now

perhaps i can make a short script that moves the mouse around and i can make a comparison video

**Comment by basro** [comment]
Well, the real question is whether the improvements we see in the mouse test are really transferring to HaxBall or not.

I don't need any more evidence that in the simple mouse test it's improving things. What I would like now is the subjective confirmation that it's better from HaxBall players. And you have already provided some of that.

I'd like for @ATTICUSHBL and @HaxPlayer1 to report on their experience comparing all the versions.

If I get no word from them by tomorrow night I'll probably just release it anyway, since you've already kindly provided some confirmation that it works. Although I'll probably have to implement a setting to disable it just in case this breaks someone's setup.

**Comment by AnddyAnddy** [comment]
> Well, the real question is whether the improvements we see in the mouse test are really transferring to HaxBall or not.
> 
> I don't need any more evidence that in the simple mouse test it's improving things. What I would like now is the subjective confirmation that it's better from HaxBall players. And you have already provided some of that.
> 
> I'd like for @ATTICUSHBL and @HaxPlayer1 to report on their experience comparing all the versions.
> 
> If I get no word from them by tomorrow night I'll probably just release it anyway, since you've already kindly provided some confirmation that it works. Although I'll probably have to implement a setting to disable it just in case this breaks someone's setup.

I've tried www and next
I'm usually an Opera GX user but here I played on Chrome 112.0.5615.138
FPS not unlocked (144)
Extrapolation 35
I also don't really know if it's a placebo or not but I THINK I noticed a better response on next
It is still better on unlocked FPS Opera GX

I also got a lower value on the desync mesuring tool, on the normal one I have around 28 but on desync it is less, I can't really exactly see how much input lag since the circle becomes too small though

**Comment by ATTICUSHBL** [comment]
> Well, the real question is whether the improvements we see in the mouse test are really transferring to HaxBall or not.
> 
> I don't need any more evidence that in the simple mouse test it's improving things. What I would like now is the subjective confirmation that it's better from HaxBall players. And you have already provided some of that.
> 
> I'd like for @ATTICUSHBL and @HaxPlayer1 to report on their experience comparing all the versions.
> 
> If I get no word from them by tomorrow night I'll probably just release it anyway, since you've already kindly provided some confirmation that it works. Although I'll probably have to implement a setting to disable it just in case this breaks someone's setup.

From my experience, in spite of being with the fps limited in 60, you can notice the difference comparing it with the current version of haxball limited in 60 fps, that is to say, the input lag is lower in the "Next" version, I did not check it with the "Old" version but regarding the current version I can say that there is a difference.

EDIT: I did the check with the "old" version and I could verify that in the "next" version you can perceive less input lag.

**Comment by kingjasko** [comment]
> Well, the real question is whether the improvements we see in the mouse test are really transferring to HaxBall or not.

I don't see why not, 3/3 people that tested it notice a positive difference

**Comment by basro** [comment]
@ATTICUSHBL that is really good news!

@AnddyAnddy it's very interesting that opera GX can run haxball with fps unlocked without freezing since opera is up to date with the more modern web features.

I've noticed that next.haxball.com wont run properly in opera with fps unlocked either. It will lock to 60 fps. Making the desynchronized canvas a video option will be sufficient probably. The problem will be communicating to users that play with unlocked fps that they now need to disable that option in the video settings...

**Comment by AnddyAnddy** [comment]
Why is the limit at 60 fps ? Why not default monitor's refresh rate (60/120/144/240...), technical limits ?
I don't really know if I feel a difference between Next Opera GX not unlocked and WWW Opera GX Unlocked regarding input lags, but I can say I can feel the difference between 144 and 60 fps

I think that as long as you provide different options within Video Settings it will be fine for us, with such a sensitive topic you should not force people to get to the newest chrome feature or whatever, placebo effect can be huge especially on haxball

**Comment by basro** [comment]
> Why is the limit at 60 fps ? Why not default monitor's refresh rate (60/120/144/240...), technical limits ?
> I don't really know if I feel a difference between Next Opera GX not unlocked and WWW Opera GX Unlocked regarding input lags, but I can say I can feel the difference between 144 and 60 fps

I have no clue why, It is not something I'm doing to actively restrict the framerate. Seems like a bug on chromium when desynchronized interacts with the fps limit disabled option.

**Comment by AnddyAnddy** [comment]
Sorry I missed something
FPS are fixed to 60 WITHOUT fps limit disabled

**Comment by basro** [comment]
@AnddyAnddy hm, ok that's really not good.

Do you also get less than 144 fps here? https://basro.github.io/input-lag-measuring-tool/desynchronized_canvas

**Comment by AnddyAnddy** [comment]
I don't know if you changed something in the meantime on Next (I restarted my pc on my side so I could have a bias) but it's fine now, I get 144 fps there and in the input lag measuring tool too WITHOUT fps limit disabled.

**Comment by basro** [comment]
> I don't know if you changed something in the meantime on Next (I restarted my pc on my side so I could have a bias) but it's fine now, I get 144 fps there and in the input lag measuring tool too WITHOUT fps limit disabled.

Is there any chance that you were mistaken and thought you were running without when it was in fact with?

**Comment by AnddyAnddy** [comment]
Not gonna lie there could be chances, I'll try to ask some people around me using Opera GX to confirm you as soon as possible, I don't guarantee to answer that soon though because it's night here

**Comment by HaxPlayer1** [comment]
@basro  Dear basro, input lag feels much less and the experience improves so much that it is even difficult for me to make a corner because of how fluid it is, I congratulate you for having fixed the input

**Comment by Round535** [comment]
Well I find myself more delayed (like with higher input lag) when certain players are on the field. And their movements are not natural (I guess they edit their clients or use gaxball website). I dont know if that is possible, but could they be ruining legit players experience (influencing our movement) by cheating them selfs?

**Comment by basro** [comment]
This feature has now been released, there's an option to disable it in the video settings.

@AnddyAnddy @kingjasko
I've found a way to work around the chrome bug where the fps is locked at 60 despite running with --disable-frame-rate-limit. 

Also, for some reason with this workaround and the low latency canvas option enabled modern chrome is no longer freezing. However, the fps it reaches seems lower: around 600 fps in my case while old chrome would reach around 1500. At the same time the power consumption seems a bit lower so this could be a good thing. Could you test how the input lag feels for you in modern chrome with the low latency canvas vs old chrome?

**Comment by kingjasko** [comment]
Well, I don't have any older versions of chromium based browser to test with, but I do notice much lower fps (fps unlock command indeed works now) on the latest version of chrome. 

Personally, I'd prefer being able to use even more of my PCs resources in exchange for much higher fps (currently cpu usage is 20% and gpu 45%), would something like that be possible? That paired with customizable fps limits in video settings would be so good for reducing input lag and increasing general performance, stability and consistency of the game.

**Comment by basro** [comment]
> Personally, I'd prefer being able to use even more of my PCs resources in exchange for much higher fps (currently cpu usage is 20% and gpu 45%), would something like that be possible? That paired with customizable fps limits in video settings would be so good for reducing input lag and increasing general performance, stability and consistency of the game.

I'm not doing anything to limit the fps so it is out of my control. If it's useful to you, it seems opera GX will also work past 60 fps now but uses as much resources as it did without the low latency canvas.

The work around I implemented was simply adding an invisible and always-on css animation. Chrome then seems to try and render that css animation and when it does it remembers to call the javascript events I need to render the game.

**Comment by AnddyAnddy** [comment]
Here are my results
# Chrome
- First I tried chrome LOCKED, 144 fps:
-- I did not feel a lot of difference between new and old canvas to be honest, I feel like both were a bit slippery, more than what i'm used to in Opera GX. It might be a bit more responsive with the new canvas

- Second I tried chrome UNLOCKED:
-- With the old canvas it is simply unplayable, it reaches 4000 fps and it is frame per frame
-- With the new canvas it definitely feels more responsive than locked chrome. I reach around 1300 fps

# Opera GX
- Third I tried Opera GX UNLOCKED (what i'm used to)
-- The old canvas feels heavily better than chrome Unlocked new canvas, I have way more control that way, around 1400 fps
-- The new canvas feels a bit more fluid than the old canvas BUT it feels less responsive like chrome unlocked new canvas, around 1800 fps

- Finally I tried Opera GX LOCKED
-- I felt the same than my first point (chrome locked)


To sum up, the best I had yesterday was Opera GX UNLOCKED and the best I have today is Opera GX UNLOCKED old canvas

If I had to rank my feelings of today:

1) Opera GX Unlocked Old canvas
2) Opera GX Unlocked New canvas
3) Chrome Unlocked New canvas
4) Opera GX Locked New/Old canvas / Chrome Locked New/Old canvas
5) (impossible to play) Chrome unlocked old canvas

**Comment by AnddyAnddy** [comment]
I think it would make it more convenient for these kinds of tests to create a Haxball Dev discord (you can pick who to invite) it would allow faster responses for both sides.

**Comment by AnddyAnddy** [comment]
@basro I think you should highlight IN GAME what the low latency canvas is doing, maybe highlight it and put a sum up of what it does when we scroll over it

Because not many people read the news and MANY people are just lost about it

**Comment by basro** [comment]
> I think it would make it more convenient for these kinds of tests to create a Haxball Dev discord (you can pick who to invite) it would allow faster responses for both sides.

Maybe someday, keeping the conversations open on github has it's advantages since others will be able to find them through google.



> @basro I think you should highlight IN GAME what the low latency canvas is doing, maybe highlight it and put a sum up of what it does when we scroll over it
> 
> Because not many people read the news and MANY people are just lost about it

Could you elaborate? What are they lost about? If there are many players for whom the update is making the game run worse please let me know.

I agree that a description of what the settings do would be useful.

**Comment by AnddyAnddy** [comment]
> > I think it would make it more convenient for these kinds of tests to create a Haxball Dev discord (you can pick who to invite) it would allow faster responses for both sides.
> 
> Maybe someday, keeping the conversations open on github has it's advantages since others will be able to find them through google.

It is true, github issues are really nice for the history, for feature requests and bug fixes. But as you are updating the game more often, maybe you would like to have fast feedback (faster than on github) on testing features.
You would keep the github for everything but whenever you need testers you would receive results faster

> > @basro I think you should highlight IN GAME what the low latency canvas is doing, maybe highlight it and put a sum up of what it does when we scroll over it
> > Because not many people read the news and MANY people are just lost about it
> 
> Could you elaborate? What are they lost about? If there are many players for whom the update is making the game run worse please let me know.
> 
> I agree that a description of what the settings do would be useful.

They don't know : 
- Whether they should activate it
- What it means 
- On which browser they should play
- If it changes something --placebo or not (mainly nocebo)--
- Some says that the performances changed even with the old canvas

I join a small suggestion:
When joining a room, maybe you could add a very short message about the purpose of the latest released update. Next to the keys to use to play and the views.
I say that because a lot of people are still using broken Haxball All in One tool, I would say 95% of them mainly (but not only) for the hide chat feature and they are not aware it is now possible directly in haxball changing the Input

**Comment by AnddyAnddy** [comment]
@basro 
I've got someone experiencing a huge drop of fps as soon as the timer hits 30 secs left and it is fine again on overtime.
It happens every time for him.

- latest version of chrome
- new canvas 
- windows 10 
- 1920x1080, 100% haxball res
- fps unlocked 
- zoom 3
- Chrome 110% zoom
- Extrapolation 85

https://user-images.githubusercontent.com/38374141/235790932-5e58b834-4135-4da6-aaf7-1989ecdf80f3.mp4

**Comment by basro** [comment]
@AnddyAnddy, my guess is that the animation is just that expensive for some reason. This is not a bug or something too surprising.

If the reduction in fps is also impacting very heavily on his input lag feel free to open an issue about it specifically. The animation is implemented through css, it's not too hard to disable it through the use of custom style sheets.
